Add trace points for socket-LB

aditighag opened the pull request:
This PR introduces tracing for socket-lb implemented in BPF cgroup hooks.

### Summary of commits:

Datapath commits - 

Commits (2-5, 7) are preparatory commits for the core changes.
**Commits 6, 8 are the core changes that introduce structures to 
send socket tracing events, and emit these traces in `bpf_sock`.**

Agent commits - 

Commits 9, 10, 11 are preparatory commits to introduce agent counterpart
of sock tracing event structures. 
Commit 12 extends Kubernetes Pod object with some more fields that are needed
to compute pod/container metadata.
**Commit 14 is the core commit that introduces a new manager for bookkeeping of 
pod and container related metadata.**
Commit 15 wires in metadata manager in daemon/watcher so that it can get callbacks
on pod watcher events.
Commit 17 adds an API for Hubble to get pod metadata so that it can associate datapath
tracing events to the corresponding pods.

### General context:

1. Overview of how Hubble will correlate datapath socket tracing events to the corresponding pods:

In packet tracing events, Hubble uses source IP parsed from the events
to get the corresponding pod metadata (e.g., name, namespace, etc). However, source
ip is not available in BPF cgroup `connect` hook as the source IP selection may happen
later in the stack based on source routing, etc. 
Hence, we introduce a new attribute "cgroup id" to uniquely identify datapath 
socket tracing events. Cgroup id is unique for every cgroup hierarchy (thereby containers and pods).

2. Why does the agent maintain metadata at the container level and not pod?

The BPF helper to get ancestor cgroup id needs an ancestor level. We can't have the
level hardcoded as the level can be different in nested environments like kind. So the
agent will need to compute it, and write it to a header file that datapath can read. There
can be unknowns around the level computation logic so we'll follow it up in follow-ups
as needed.

3. How does the agent get cgroup path for containers?

In order to get containers' cgroup ids, the agent first needs to get
their cgroup paths. We considered 3 approaches to get this low-level 
metadata that's not available in k8s watcher events:
- Compute cgroup path based on pattern (<cgroup-root>/kubepods/pod<id>/containerId (selected) 
- Use crictl CLI tool to get the cgroup path metadata. The tool isn't guaranteed to be present
   on minimalistic distros, also the API can change without being backwards comptabile.
- Implement gRPC client to communicate to underlying container runtimes over CRI (container
  runtime interface). This requires the agent to have access to the container runtime endpoint 
  socket, which can have security related risks.

WIP:
- [x] Extend cilium status with the feature flag

Follow-ups: https://github.com/cilium/cilium/issues/21510

Out of scope/Follow-ups:
- Flow validation tests based on Hubble
- Hubble receives datapath events, but the agent doesn't have container related metadata available? There are ways to address this, but we'll do it as a follow-up.
- Differentiating from host netns socket events from processes (this is doable by ignoring events for host cgroup id)
- Monitor aggregation (e.g., filter `connect` traces)
- CLI to inspect the cgroup manager internal state

**Release note**
```release-note
Add tracing for socket-based load balancing.
```

Fixes: https://github.com/cilium/cilium/issues/19623

aditighag committed:
daemon,pkg/option: Introduce config for socket traces

The config will be used to enable/disable socket
traces when socket-lb is enabled, independent of
packet related traces.
Per packet traces can produce significant number
of traces, and could lead to complexity issues,
so it's good to keep the two configs separate.

The socket tracing config will be disabled when
user disables it, socket-lb is disabled, or
the underlying kernel doesn't support the feature.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Add header file to define socket context

This is a preparatory commit to define socket
related context. This context will currently
be used for socket tracing events that are
emitted from `bpf_sock`.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Define helper to send perf events for socket context

This is a preparatory commit that defines a
helper to send traces via perf event ring buffer with
socket context.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Define structures and helpers to send socket traces

This commit defines structures and helpers to be able
to send socket traces from BPF cgroup hooks. This enables
users to debug/track socket based load-balancing events.

Highlights of some fields sent in the trace notifications:
- Destination IP address/port pre- and post- service translation
- Socket cookie: This is a unique identifier for a socket
  that can be used to associate multiple trace events
  generated for a socket.
- Cgroup id: This field is unique to every cgroup hierarchy (thereby
  every container). It'll be used by the agent to correlate socket
  trace events to the corresponding parent pods of the containers.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Add trace point for socket traces

The newly added trace point will be used
in follow-up commits to send traces from
cgroup hooks executed in `bpf_sock` that implements
socket-based load balancing.
A new trace point is used rather than using
the existing `CILIUM_NOTIFY_TRACE` point as
the userspace parsers are going to be completely
independent. Socket traces will not have packet header/data
that per packet traces carry.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Refactor common code

This is a preparatory commit to
move common code so that it can be reused.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/monitor: Add message type for socket traces

The new type will be used to parse socket trace
events from the datapath in case of socket-lb.
A new trace type is used rather than using
the existing `MessageTypeTrace` point as
the userspace parsers are going to be completely
independent. Socket traces will not have packet header/data
that per packet trace events carry.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/monitor: Add struct definition for socket trace message

This is a counterpart of the `TraceSockNotify` struct
used in the datapath to send socket tracing events
to userspace.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/k8s: Extend pod type

Extend the pod type with container status
with containerID, and qos fields. This metadata
will be used to compute the cgroup path for
pods and containers.

The current use case is to correleate datapath
socket-lb trace events to the corresponding pods.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

gandro commented on pkg/monitor/datapath_sock_trace.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+// SockTraceNotify is message format for socket trace notifications from datapath.
+type SockTraceNotify struct {
```
Can we call this TraceSockNotify? It seems to be called that everywhere else.

gandro commented on pkg/monitor/datapath_sock_trace.go:
```diff
@@ -0,0 +1,30 @@
+	DstPort    uint16
+	CgroupId   uint64
+	SockCookie uint64
+	Flags      uint8
```
Let's also define some bitmaks in Go to allow user space to extract the flags.

aditighag replied:
Can we address this in the parser changes?

gandro reviewed (commented):
Looking good overall. Some preliminary feedback that I noticed while trying to write a Hubble parser.

gandro commented on bpf/lib/trace_sock.h:
```diff
@@ -0,0 +1,120 @@
+
+/* Direction for translation between service and backend IP */
+enum xlate_point {
+	XLATE_PRE_DIRECTION_FWD = 0,  /* Pre service translation */
```
Can we start this at offset 1? Similarly to `L4_PROTOCOL_UNKNOWN`. Having a zero-value for "UNKNOWN" simplifies the Hubble API, as we can assume value 0 for everything which is not source from `trace_sock_notify`.

aditighag committed:
pkg/metadata: Add pod/container cgroup metadata manager

The manager is responsible for bookkeeping of pod
and container related low-level metadata such as
cgroup path and cgroup id along with kubernetes
metadata, that'll be used for features like socket-lb tracing.

To that end, the cgroup manager:
- Implements callback functions for pod update/delete watcher events
- Processes pod events, and updates internal caches with pod
  and container related metadata, including low-level metadata
  such as cgroup path and cgroup id.
- Exposes API to get pod metadata for a container cgroup id.
  This API will be invoked by Hubble in follow-up commit in order
  to augment the datapath socket tracing events with pod metadata.

Also, added unit tests to validate pod processing events.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
daemon, pkg/k8s: Wire in cgroup metadata manager

Wire the cgroup manager in watcher via
the daemon. This will allow the watcher to
invoke cgroup manager's callbacks on pod
add/update/delete events, and the manager
will update/delete pod metadata, currently
used for socket-lb tracing.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/k8s: Invoke cgroup manager callbacks for pod events

The cgroup manager maintains low-level metadata such as
cgroup ids along with Kubernetes related metadata for pods
and containers. This is currently used for socket-lb tracing.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/hubble: Add pod metadata getter

Hubble needs to correlate datapath socket
tracing events to the corresponding pods based
on the container cgroup id emitted in the trace
events. This allows Hubble to add user-friendly
metadata (e.g., pod ips) to the socket tracing events in Hubble.
Add a new pod metadata getter that's implemented by
the metadata manager implemented in previous commits.

Follow-up changes in Hubble will invoke this API.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Emit traces in bpf_sock

Use the socket tracing helpers introduced in previous commits
to send traces from the cgroup hooks that are executed on
connect and send/receive socket calls.

Traces will be sent in the forward as well as reverse path of
service translation performed as part of socket load-balancing.
Additionally, send them pre- and post- service translation in
order to report service clusterIP address and selected backend
IP address, respectively.
The cgroup id allows the agent to correlate these events to the
corresponding pod. While the socket cookie passed in the events
allows the agent to tie together multiple events associated with
a socket (e.g., forward and reverse service translation).

We are currently sending cgroup id of containers in the trace notifications,
while Hubble mainly needs to attribute trace events at the pod level.
In future, we can use the `bpf_get_current_ancestor_cgroup_id` helper
to send pod level cgroup id directly, but the helper needs an ancestor
level that can vary in nested environments like kind.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Check for additional BPF helpers

The `bpf_get_current_cgroup_id` will be
used in follow-up commits to get cgroup
id of the container cgroup for socket-lb
tracing.


Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/cgroups: Helper to get cgroup id for cgroup path

Add a new helper that invokes the `name_to_handle_at` [1]
syscall to get the cgroup id for a cgroup path.
This will be used for socket-lb tracing feature to map
datapath socket-lb events to the corresponding pods in
the agent.

[1] https://man7.org/linux/man-pages/man2/name_to_handle_at.2.html

Signed-off-by: Aditi Ghag <aditi@cilium.io>

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+	}
+}
+
+func (m *Manager) processPodEvents() {
```
I think that there is a lot of overlap between the core structure here and the core implementation of `EventQueue`, including handling responses to enqueued events and shutdown. Beyond that `EventQueue` also has support to drain the queue on shutdown (which can signal failures to anyone blocking for the response to processing the event), and also spanstats so we can analyze when an EventQueue gets stuck, how much time is spent enqueueing/processing/etc. It might be a relatively small lift to switch this code over to use the same EventQueue implementation.

aditighag replied:
I took a look at `EventQueue` implementation, and some of the consumers. The impression I got is that the queue is particular useful for operations like draining events, and tracking time durations waiting for an event to be enqueued (e.g., the endpoint generation). At this point, I'm not seeing either of these being problematic in the metadata manager - the event processing handlers are small, and I'm not sure what the draining logic will look like as we don't have any caller contexts to cancel, or things to roll back.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metadata")
```
I think that naming this the "metadata manager" is likely to cause some confusion, just given how many different types of metadata we deal with and where it all gets stored. If this is primarily related to the way that Cilium Endpoints / pods relate to Cgroups, then maybe it'd be less confusing to keep the name a bit more concrete, perhaps by having it live under `pkg/cgroups/manager/` and referring to this as CgroupManager?

aditighag replied:
Yeah, I was debating about the name. Thanks for the suggestion - I like renaming it to `CgroupManager`. 
I renamed the manager itself, but left it under `pkg/metadata` as there is k8s metadata involved as well, and `pkg/cgroups` mainly relates to cgroup related operations.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type Manager struct {
+	// Map of pod metadata indexed by their ids
+	podMetadataById map[string]*podMetadata
```
What is the source of this ID? Can we use a typed ID here so that it's easy to trace the exact type of this ID / which parts of the code generate it?

Maybe it's just enough to update the comment to mention that this is the Pod UID.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+	m := &Manager{
+		podMetadataById:           make(map[string]*podMetadata),
+		containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+		podEvents:                 make(chan podEvent, 20),
```
The channel size here should be a configurable parameter on the commandline (even if it's hidden). Ideally during testing we can even set it to be very low (perhaps even serial) in order to test for deadlocks in the case that there are too many events in the system to handle correctly.

joestringer replied:
^^ Looking over the implementation the risk of deadlock looks pretty low to me, since this code isn't calling out to the rest of the daemon, but I think we should be trying to move towards having configurable channel sizes and testing with low channel sizes so that we aren't relying on having large empty channel buffers just to ensure progress is made in the agent. I wonder whether we should just have a hidden generic channel size flag for defaults on stuff like this or whether each channel should get independent flags :thinking:

aditighag replied:
Hmm... Right, there is no risk of deadlock as all the calls are serialized via the channel, and there are no locks (it's been a conscious decision so far). The buffer size merely allows callers to not block on this channel. 

> I wonder whether we should just have a hidden generic channel size flag for defaults on stuff like this or whether each channel should get independent flags

Interesting point, although in this particular case when would you need to change the size if we are not worried about deadlocks? In general, I could see some value in setting channel buffer size to a low value in unit tests.

joestringer replied:
>  I could see some value in setting channel buffer size to a low value in unit tests.

Yeah, this point is the main thing I'm thinking about. If we reduce all of these channels to a small value, then we can ensure that the system still makes progress and doesn't livelock/deadlock if there are too many events going on.

aditighag replied:
So I updated the channel size in unit tests as discussed above. Unbuffered channel in unit tests should catch any issues on the caller side. Good suggestion! Thanks!

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+// During initialization, the manager checks for a valid base cgroup path.
+// In case of environments using non-default cgroup base paths, manager will
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type Manager struct {
```
If I follow correctly, the internals of this structure *must* be accessed only by the main goroutine inside the manager, and that's how we keep it lockless (basically by implementing message passing via podEvents for any/all interactions). It's probably worth mentioning this in a comment above the structure here as a hint to future reviewers that other goroutines must not start interacting with these fields.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+	m.cgroupsChecker = cg
+	m.checkCgroupPath = new(sync.Once)
+
+	m.enable()
```
I wonder if we could get away without the `m.enabled` bool by creating an interface for the cgroups manager that just exposes the external-facing API, then create a dummy one that's initialized in the case when `TraceSockNotify` is disabled. :thinking: 

The tricky part is that it looks like `TraceSockNotify` can be dynamically enabled/disabled at the moment, so we'd need a way to replace that on-the-fly :thinking:

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+	// Object to check cgroup base path
+	checkCgroupPath *sync.Once
+	// Flag to check if manager is enabled, and processing events
+	enabled atomic.Bool
```
Why does enabled need to be atomic? It looks like it's always set true from the caller of NewManager and then that goroutine starts the other goroutine which can check the bool.

Did you plan to make this able to be dynamically enabled/disabled?

aditighag replied:
So the flag is enabled depending on whether socket-lb is enabled or not. I didn't initially realize that I had added the flag in the list that's enabled dynamically at run time. Anyway the latest push reflects the right intentions. I think this also addresses your above comment.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+		podId, containerId), nil
+}
+
+func newManagerTest(fs fs, cg cgroup) *Manager {
```
nit: Does this test code need to be in the main file? Seems like there could perhaps be a little more sharing with the main NewManager implementation and/or shift this test one that passes in dummy `fs` and `cg` implementations into the test file itself.

joestringer reviewed (commented):
Nice work! I looked primarily over pkg/metadata so far and the overall structure makes a lot of sense to me. I think there's a couple of opportunities for more code reuse and also one open question around how dynamic enable/disable of the socket trace notifications will react (ie `cilium config TraceSockNotification=Enabled` / Disabled). I didn't dig too deep on other stuff like log message usage and so on yet since there's still some open TODOs in the summary.

gandro commented on pkg/monitor/datapath_sock_trace.go:
```diff
@@ -0,0 +1,32 @@
+
+// Socket trace event point with respect to service translation
+const (
+	XlatePointPreDirectionFwd = iota
```
This is now off by one:

```suggestion
	XlatePointUnknown = iota
	XlatePointPreDirectionFwd
```

gandro commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,404 @@
+	name      string
+	namespace string
+	ips       []v1.PodIP
+}
```
Those fields need to be exported (or this type needs to provide a getter) for `GetParentPodMetadata` to be useful outside of this package.

aditighag committed:
agent: Extend cilium status with socket lb tracing feature

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -312,6 +315,13 @@ func initKubeProxyReplacementOptions() (bool, error) {
 			option.Config.EnableSocketLB = false
+			option.Config.EnableSocketLBTracing = false
+		}
+		option.Config.EnableSocketLBTracing = true
```
I've updated this block to fix the control plane test issue. 

```
		option.Config.EnableSocketLBTracing = true
		if !option.Config.DryMode {
			found := false
			if h := probesManager.GetHelpers("cgroup_sock_addr"); h != nil {
				_, found = h["bpf_perf_event_output"]
			}
			if !found {
				option.Config.EnableSocketLBTracing = false
			}
		}
```

borkmann commented on bpf/lib/trace_sock.h:
```diff
@@ -0,0 +1,150 @@
+		.sock_cookie = sock_cookie,
+		.cgroup_id = cgroup_id,
+		.l4_proto = parse_protocol(l4_proto),
+		.ipv6 = 0,
```
small nit, please align all `=`, so it's easier for readers. ditto below

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -1135,8 +1140,9 @@ static __always_inline int __sock6_xlate_rev(struct bpf_sock_addr *ctx)
 	struct ipv6_revnat_tuple key = {};
 	struct ipv6_revnat_entry *val;
 
+	__u16 dst_port = ctx_dst_port(ctx);
```
small nit: please use kernel coding style

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -357,6 +362,15 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
+	send_trace_sock_notify4(ctx_full, XLATE_PRE_DIRECTION_FWD, dst_ip,
+				bpf_ntohs(dst_port), sock_cookie, cgroup_id,
+				ctx_full->protocol);
+#endif
```
Bit too much ifdef all around the code here. Please hide sock cookie and cgroup_id retrival inside send_trace_sock_notify{4,6}, so this becomes just a simple function call in the main code.

Also why the double BPF_HAVE_PERF_EVENT_OUTPUT && TRACE_SOCK_NOTIFY?

Can't the agent probe for the helper and only then if available enable TRACE_SOCK_NOTIFY so you can simplify all the ifdefs?

Same with BPF_HAVE_CGROUP_ID, I presume the thought is to still support this infra even when get_current_cgroup_id is not available, right (instead of making it dependent on it)?

nit: Do we need the extra ctx_full->protocol, can't we retrieve it from inside send_trace_sock_notify{4,6} when we pass th ctx_full in there?

Otherwise rest of bpf/ looks good though.

kaworu commented on pkg/hubble/parser/getters/getters.go:
```diff
@@ -76,3 +77,10 @@ type LinkGetter interface {
 	Name(ifIndex uint32) string
 }
+
+// PodMetadataGetter ...
```
nit: I know it's the case for most of the getter above in this file, but I think we should either document this properly or drop the comment.

aditighag replied:
This particular getter warranted some description, so I added one. :)

kaworu reviewed (approved):
Hubbles changes LGTM.

aditighag commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+}
+
+func (m *CgroupManager) enable() {
+	go m.processPodEvents()
```
I've made a small change locally to start the goroutine only after the `m.enabled` flag is set, and move the common code to `initManager(...)`. So both manager and manager_test call into the `initManager(...)`. 

<details><summary>diff</summary>
<p>

diff --git a/pkg/metadata/manager.go b/pkg/metadata/manager.go
index 38a9c2d8e1..c122c2c7df 100644
--- a/pkg/metadata/manager.go
+++ b/pkg/metadata/manager.go
@@ -73,19 +73,7 @@ const (
 )

 func NewCgroupManager() *CgroupManager {
-       m := &CgroupManager{
-               podMetadataById:           make(map[string]*podMetadata),
-               containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
-               podEvents:                 make(chan podEvent, podEventsChannelSize),
-               shutdown:                  make(chan struct{}),
-       }
-       m.fschecker = fsImpl{}
-       m.cgroupsChecker = cgroupImpl{}
-       m.checkCgroupPath = new(sync.Once)
-
-       m.enable()
-
-       return m
+       return initManager(fsImpl{}, cgroupImpl{}, podEventsChannelSize)
 }

 func (m *CgroupManager) OnAddPod(pod *v1.Pod) {
@@ -188,8 +176,24 @@ func (c cgroupImpl) GetCgroupID(cgroupPath string) (uint64, error) {
        return cgroups.GetCgroupID(cgroupPath)
 }

+func initManager(fs fs, cg cgroup, channelSize int) *CgroupManager {
+       m := &CgroupManager{
+               podMetadataById:           make(map[string]*podMetadata),
+               containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+               podEvents:                 make(chan podEvent, channelSize),
+               shutdown:                  make(chan struct{}),
+       }
+       m.fschecker = fs
+       m.cgroupsChecker = cg
+       m.checkCgroupPath = new(sync.Once)
+
+       m.enable()
        go m.processPodEvents()
+
+       return m
+}
+
+func (m *CgroupManager) enable() {
- go m.processPodEvents()
</p>
</details>

joestringer commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -62,15 +62,17 @@ func initKubeProxyReplacementOptions() (bool, error) {
+		option.Config.EnableSocketLBTracing = false
 		option.Config.EnableHostServicesTCP = false
 		option.Config.EnableHostServicesUDP = false
+		option.Config.EnableSocketLBTracing = false
```
Duplicate?

joestringer commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -312,6 +315,13 @@ func initKubeProxyReplacementOptions() (bool, error) {
+		option.Config.EnableSocketLBTracing = true
+		if h := probesManager.GetHelpers("cgroup_sock_addr"); h != nil {
+			if _, found := h["bpf_perf_event_output"]; !found {
+				option.Config.EnableSocketLBTracing = false
```
I think this should be accompanied with a warning if this overrides the user's configuration? Otherwise it will just silently turn off if the kernel support isn't there.

joestringer commented on pkg/monitor/format/format.go:
```diff
@@ -249,6 +251,9 @@ func (m *MonitorFormatter) FormatSample(data []byte, cpu int) {
 	case monitorAPI.MessageTypeRecCapture:
 		m.recorderCaptureEvents(prefix, data)
+	case monitorAPI.MessageTypeTraceSock:
+		log.Infof("sock-traces-debug %+v %v", data, messageType)
```
This should not print frequent events to the log at `Info` level.

aditighag replied:
This was just a stub commit, I've removed it. Sebastian will be adding a parser to dump the data in follow-ups.

joestringer commented on pkg/monitor/format/format.go:
```diff
@@ -249,6 +251,9 @@ func (m *MonitorFormatter) FormatSample(data []byte, cpu int) {
 		m.recorderCaptureEvents(prefix, data)
+	case monitorAPI.MessageTypeTraceSock:
+		log.Infof("sock-traces-debug %+v %v", data, messageType)
+		fmt.Printf("%s Sock event: %+v\n", prefix, data)
```
I think the `data` probably needs to be properly formatted somehow? Looks like it's just `[]byte` at this point, which will not be very user-friendly.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+package metadata
```
Missing license/SPDX preamble.

joestringer commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+package metadata
+
```
Does it still make sense to put this in `pkg/metadata/`? Maybe it should belong under `pkg/cgroups/` instead?

aditighag replied:
Moved it under `pkg/cgroups/manager` as you had originally suggested.

joestringer reviewed (commented):
I skipped the datapath, k8s and the core `CgroupManager` logic this time but glanced over the other bits. I saw a couple of comments you left that you have local changes, so figured I should wait before taking a closer look at the core logic again.

squeed commented on pkg/hubble/parser/getters/getters.go:
```diff
@@ -76,3 +77,10 @@ type LinkGetter interface {
 }
+
+// PodMetadataGetter ...
+type PodMetadataGetter interface {
```
The term "Parent Pod" is a bit confusing; a Container is part of a Pod by default. ParentPod somehow implies the ancestor of another pod. Would it make sense to rename this?

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+	IPs       []string
+}
+
+const (
```
Can you add a quick docblock?

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+	PodAddEvent = iota
+	PodUpdateEvent
+	PodDeleteEvent
+	PodMetadataEvent
```
Nit: should this be called GetPodMetadata or something?

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+			log.Debugf("failed to get cgroup id for cgroup path (%s): %v", cgrpPath, err)
+			continue
+		}
+		log.Infof("aditi-debug-metadata %s %d", cgrpPath, cgrpId)
```
Mean to delete this?

aditighag replied:
Yes :) Missed deleting the debug log.

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+
+// GetParentPodMetadata returns parent pod metadata for the given container
+// cgroup id in case of success, or nil otherwise.
+func (m *CgroupManager) GetParentPodMetadata(cgroupId uint64) *PodMetadata {
```
Is there a reason you didn't want to read from the map directly (behind a sync.RWMutex)? Seems like it might be easier.

aditighag replied:
I made a conscious choice to process all the events via a channel, and didn't want to deal with locks at this time. The channel approach is more easy to follow, and no risk of forgetting to lock, or run into deadlocks, etc.

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+	currContainers := make(map[string]struct{}, numContainers)
+	for _, c := range pod.Status.ContainerStatuses {
+		var cId string
+		if cId = c.ContainerID; cId == "" || c.State.Running == nil {
```
If one container in a pod crashes and is not restarted, should we delete it from the store?

squeed replied:
Nevermind :-)

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+	}
+	// Clean up any pod's old containers.
+	if oldPod != nil {
+		for _, c := range oldPod.Status.ContainerStatuses {
```
why not `range pm.containers` here?

aditighag replied:
Well, we first update `pm.containers` based on the latest state, and then clean up from the old pod containers. Didn't seem worthwhile to first do a diff.

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+		// Container could've been gone, so don't log any errors.
+		cgrpPath, err := m.getContainerCgroupPath(id, cId, pod.Status.QOSClass)
+		if err != nil {
+			log.Debugf("failed to get container metadata for (%s): %v", cId, err)
```
suggestion: always log pod namespace + name (whenever possible)

squeed replied:
This goes for most of the log lines in this module.

aditighag replied:
Good point. I'll update the logs with pod namespace.

squeed commented on pkg/metadata/manager.go:
```diff
@@ -0,0 +1,374 @@
+		return "", fmt.Errorf("failed to get cgroup path for (%s)", containerId)
+	}
+
+	return fmt.Sprintf("%s/pod%s/%s", m.baseCgroupPathForQos(m.templateCgroupBasePath, containerQos),
```
Is this guaranteed to always be the case? Even for cri-o?

aditighag replied:
We are using known defaults that are also hard-coded in Kubernetes source code. What about cri-o -  is the base cgroup path different?

squeed replied:
I'm not sure, but the path does look different. Looking at a random container in a pod, I see this path (on the host):

`/sys/fs/cgroup/systemd/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podaada424e_b978_437b_bd12_5e22626a3c08.slice/crio-9b6be68b035f1b3453b0a0deca929d6c1953ce9bc9b4f18e6a44a9d8868e0328.scope`

aditighag replied:
Can you post output of `sudo crictl inspect <container-id> | grep cgroupsPath`? If you have a code reference for Kubelet, then it'll be easy to adapt the logic in case crio has a different cgroup path.

squeed commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -257,6 +257,7 @@ cilium-agent [flags]
       --tofqdns-proxy-port int                                  Global port on which the in-agent DNS proxy should listen. Default 0 is a OS-assigned port.
       --tofqdns-proxy-response-max-delay duration               The maximum time the DNS proxy holds an allowed DNS response before sending it along. Responses are sent as soon as the datapath is updated with the new IP information. (default 100ms)
       --trace-payloadlen int                                    Length of payload to capture when tracing (default 128)
+      --trace-sock                                              Enable tracing for socket-based LB
```
Is there a reason *not* to enable this?

aditighag committed:
bpf, monitor: Add alignment checks for trace sock structs

Signed-off-by: Aditi Ghag <aditi@cilium.io>

nathanjsweet reviewed (approved):
LGTM for API.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+			}).Debugf("failed to get cgroup id for cgroup path (%s): %v", cgrpPath, err)
+			continue
+		}
+		log.Infof("aditi-debug-metadata %s %d", cgrpPath, cgrpId)
```
```suggestion
```

aditighag replied:
Aye, already removed it locally.

joestringer commented on pkg/cgroups/cgroups_linux.go:
```diff
@@ -54,3 +56,15 @@ func cgrpCheckOrMountLocation(cgroupRoot string) error {
+
+func GetCgroupID(cgroupPath string) (uint64, error) {
+	var handle unix.FileHandle
+	handle, _, err := unix.NameToHandleAt(unix.AT_FDCWD, cgroupPath, 0)
```
```suggestion
	handle, _, err := unix.NameToHandleAt(unix.AT_FDCWD, cgroupPath, 0)
```

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+			}).Errorf("unexpected container ID: %s", cId)
```
Might be good to include the `c.Name` here too if we have it?

aditighag replied:
`c.Name` isn't available.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+			// is the same.
+			continue
+		}
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
```
Should these wait until after the `continue` statements below since we could end up skipping this container if we can't locate the cgroup path or id from that path?

aditighag replied:
Cgroup path or id don't change for a container. So my thinking is that we don't want to repeat the processing. Wdyt?

joestringer replied:
This question is maybe a bit more theoretical: broadly there are "error" cases in the "continue" statements below which could cause this container to be skipped. In that case, `m.podMetadataById[id]` would be updated via `pm.containers[cId]` here, but this loop wouldn't reach the `m.containerMetadataByCgrpId[cgrpId]` update below on line 323. Just as a consistency thing, I would think that the structure should ideally have both of these updated or neither.

aditighag replied:
>In that case, m.podMetadataById[id] would be updated via pm.containers[cId] here, but this loop wouldn't reach the m.containerMetadataByCgrpId[cgrpId] update below on line 323. 

The error condition is that we weren't able to compute metadata for pod's containers. This can be useful for debugging when we add a CLI in future.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+	if qos == v1.PodQOSGuaranteed {
+		return fmt.Sprintf("%s", path)
+	}
+	return fmt.Sprintf("%s/%s", path, strings.ToLower(string(qos)))
```
Are these filesystem paths? Should we be using [filepath](https://pkg.go.dev/path/filepath).Join() instead?

aditighag replied:
Sure, might as well use `filepath.Join` instead.

joestringer commented on pkg/k8s/watchers/pod.go:
```diff
@@ -229,6 +229,8 @@ func (k *K8sWatcher) addK8sPodV1(pod *slim_corev1.Pod) error {
 		}
 	}
 
+	k.cgroupManager.OnAddPod(pod)
```
Should these calls be protected by a check for `option.Config.EnableSocketLBTracing`?

At least today only the socket LB tracing will use the cgroup manager. This may of course change in future, but I think it will help if this code is conditional on the feature flags that uses it.

(Same question for the initialization in the daemon, but maybe it's easier to just unconditionally initialize the object in the daemon for now and then maybe we can factor this out as a `pkg/hive:Cell` module as that framework matures)

joestringer replied:
Given that we are expecting to extend this to be used by other features soon, it's probably not worth adding conditional checks here. We can handle this in the cgroup manager internals.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+func (m *CgroupManager) processPodEvents() {
+	for {
+		select {
+		case ev := <-m.podEvents:
+			if !m.enabled {
+				continue
+			}
```
Ah I see, this is maybe a partial answer to my question above: Rather than scattering conditional checks for whether SocketLBTracing should be enabled across the codebase, this always runs a goroutine and just skips over the events if the flag is disabled?

I think the ideal model would be that we never initialize this code and never have it subscribe to the watcher events if the flags are disabled. Maybe for the short term it doesn't really matter, but it's a bit of an odd structure to have the watcher send events on this channel only for this channel to ignore the event notification.

(Practically I don't think there's likely to be that much impact with the frequency of events we're dealing with here, but I look forward to when we can refactor the k8s watchers to allow modules like this to just subscribe to events and not force the watcher to know about how to call out to these modules in order for these notifications to be processed)

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+		}
+		if m.templateCgroupBasePath == "" {
+			log.Warnf("No valid cgroup base path found: socket " +
+				"load-balancing tracing feature will not work")
```
Couple of notes here:
- No need to use the formatting variation of the logging function if there is no formatting directives
- Should we suggest to the user to collect some specific information from their system and file an issue on github? If they hit this, we will want to know what the real cgroup path is.
```suggestion
			log.Warn("No valid cgroup base path found: socket " +
				"load-balancing tracing feature will not work")
```

aditighag replied:
>Should we suggest to the user to collect some specific information from their system and file an issue on github? If they hit this, we will want to know what the real cgroup path is.

Yeah, I was going to put this in documentation so that we can mention what the current defaults are. But having it logged as part of the warning is fine too, we can add more details to the documentation.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+			}).Errorf("unexpected container ID: %s", cId)
```
Please also parameterize the `cId` here. It should not be part of the error string, it should be a field like the other fields.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+			}).Debugf("failed to get container metadata for (%s): %v", cId, err)
+			continue
+		}
+		cgrpId, err := m.cgroupsChecker.GetCgroupID(cgrpPath)
```
If we ever get restricted on the underlying syscalls/fs access by something like SELinux here, I think we're going to wonder why the feature silently doesn't work :thinking: 

Not sure what the best option is to handle this. I agree we don't want to be printing errors all the time for expected conditions, but also if something slightly unusual is going on then this is going to break silently and it's going to be difficult to debug. Hmm.

aditighag replied:
Hmm... Not sure what the best way to handle this either. Perhaps use the cilium pod as a reference, and log a one time error?

joestringer replied:
:bulb: ah interesting, yeah so just do a one-off detection on startup that should be a sign of whether this call would work, and emit an error in that case since that should be enough to indicate whether there is likely ever a problem with this or not. :+1:

aditighag replied:
Cilium pod might not be the very first pod that the cgroup manager receives events for. I've added a TODO in the code to revisit this as there are unknowns such as what would the errors even look like.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,392 @@
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type CgroupManager struct {
+	// Map of pod metadata indexed by their UIDs
+	podMetadataById map[podUID]*podMetadata
```
I feel like we are going to want a `cilium cgroups list` CLI command to introspect this, if there's ever trouble while trying to map the socket messages to pods/containers. Did you ever find yourself needing something like this during development?

Might be an interesting followup.

aditighag replied:
Good feedback! Having a CLI command to inspect the internals sounds handy for debugging. Added to the follow-ups.

aspsk commented on bpf/lib/trace_sock.h:
```diff
@@ -0,0 +1,149 @@
+}
+
+static __always_inline void
+send_trace_sock_notify4(struct __ctx_sock *ctx __maybe_unused,
```
All these arguments are actually used

aspsk commented on bpf/lib/trace_sock.h:
```diff
@@ -0,0 +1,149 @@
+	struct trace_sock_notify msg __align_stack_8;
+	__u64 cgroup_id = 0;
+	__u64 sock_cookie;
+	__u32 l4_proto = ctx->protocol;
```
Do we follow the kernel style in our bpf code? If so, then this should be moved upper.

Do you really need the `l4_proto` and `sock_cookie` variables in this function? Without them it could be 3 lines less without affecting readability (IMHO).

aspsk commented on bpf/bpf_sock.c:
```diff
@@ -328,9 +298,11 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
-		.address	= ctx->user_ip4,
-		.dport		= ctx_dst_port(ctx),
+		.address	= dst_ip,
+		.dport		= dst_port,
```
I wonder, why this and similar changes below were needed?

aditighag replied:
This was just to reuse the variables.

aspsk replied:
Yeah, but they aren't reused :) Anyway, I am ok with this change, maybe it looks better indeed.

aditighag replied:
Both `dst_ip` and `dst_port` are reused when passed to `send_trace_sock_notify`. See one of the previous refactoring commits that moved the value to variables.

aspsk replied:
Ok, I see, it's a later commit (I read changes by commits). Thanks!

aspsk commented on bpf/include/bpf/helpers.h:
```diff
@@ -104,4 +104,6 @@ static int BPF_FUNC_REMAP(get_socket_opt, void *ctx, int level, int optname,
 			  void *optval, int optlen) =
 	(void *)BPF_FUNC_getsockopt;
 
+static __u64 BPF_FUNC(get_current_cgroup_id);
```
This should be defined two commits earlier, when you start using it in `bpf/lib/trace_sock.h`, otherwise, the per-commit compilation may fail (it doesn't fail in tests because `TRACE_SOCK_NOTIFY` is not set)

aditighag replied:
Good catch! Fixed. Thanks!

aspsk reviewed (changes requested):
Looked at the datapath part, looks great. One minor fix (move a definition to a different commit), and some style comments (which I am not insist on, more wondering).

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -414,6 +418,10 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
+
+	send_trace_sock_notify4(ctx_full, XLATE_POST_DIRECTION_FWD, backend->address,
+				bpf_ntohs(backend->port));
+
```
From trace PoV, what is the rationale for XLATE_PRE_DIRECTION_FWD and XLATE_POST_DIRECTION_FWD compared to just having a XLATE_DIRECTION_FWD hook which is done after each point where we rewrite ctx->user_ip and port?

Meaning, adding XLATE_DIRECTION_FWD with svc vip/port plus backend ip/port info that we pass to send_trace_sock_notify4() below before the return 0, and also for the L7 case where we write (which seems missing here?).

And then having XLATE_ERROR_... tracepoints where we get more visibility in some of the error cases which are proper errors, like failing to update sock4_update_revnat etc. Wdyt?

aditighag replied:
Discussed offline, posting here for posterity.

The intent behind having pre and post service translation trace points is to enable users to track the logic separately. For early returns, we already have error metrics for visibility. 

This is the note for L7, where it skips socket-lb explicitly:
```
/* Do not perform service translation at socker layer for
	 * services with L7 load balancing as we need to postpone
	 * policy enforcement to take place after l7 load balancer and
	 * we can't currently do that from the socket layer.
	 */
```

The error codes are out of the scope of the PR, but I'm happy to discuss this so that we can extend this in follow-ups if/as needed.

aditighag committed:
monitor: Add switch case for trace sock messages

The events will be formatted to show user-friendly
data.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aspsk commented on bpf/lib/trace_sock.h:
```diff
@@ -0,0 +1,143 @@
+	if (is_defined(BPF_HAVE_CGROUP_ID))
+		cgroup_id = get_current_cgroup_id();
+	msg = (typeof(msg)){
+		.type			= CILIUM_NOTIFY_TRACE_SOCK,
```
Indentation broke a bit

aditighag replied:
This seems perfectly aligned in my IDE. Something is overriding it, it seems.

aditighag replied:
Fixed it using vim. :) IDE is acting up.

aspsk reviewed (approved):
Some indentation fixes, otherwise LGTM

joestringer commented on pkg/monitor/format/format.go:
```diff
@@ -249,6 +249,8 @@ func (m *MonitorFormatter) FormatSample(data []byte, cpu int) {
 	case monitorAPI.MessageTypeRecCapture:
 		m.recorderCaptureEvents(prefix, data)
+	case monitorAPI.MessageTypeTraceSock:
+		// TODO: events to be formatted
```
Is this for outputting in `cilium monitor`? I wonder if it's possible to implement this with just a few lines of code and be able to view these events?

aditighag replied:
PR to add a monitor parser - https://github.com/cilium/cilium/pull/21516.

joestringer reviewed (approved):
Userspace side LGTM. Just one comment wondering about how much more work it is to be able to somehow view these events, but we can also follow up on that item in a separate PR.

joestringer commented on pkg/cgroups/manager/manager.go:
```diff
@@ -0,0 +1,405 @@
+	podMetadata.IPs = append(podMetadata.IPs, pm.ips...)
+	log.WithFields(logrus.Fields{
+		"container-cgroup-id": cgroupId,
+	}).Debugf("Pod metadata: %+v", podMetadata)
```
Hey @aditighag how frequently would you expect this log to be triggered? I'm noticing it 1000+ times in the first 5 minutes of Cilium running, which seems a bit high. Given the channel usage and so on this might be introducing more memory churn / pressure than necessary. Is it expected? Many of the logs seem to repeat for the exact same Pods over and again.

aditighag replied:
This log was added before we had visibility via CLI (https://github.com/cilium/cilium/pull/23641). We can remove the log now.

joestringer replied:
My question is less about the log itself (though hiding that would reduce noise in debug mode), but more about whether we expect this function to be called so often and whether that indicates a structural problem particularly if we're creating and destroying channels all the time. Is this intended as a cheap cache that Cilium should be calling all the time or is expected to be more expensive and shouldn't be called so often?

I guess it's a bit hard to tell with my report given that I don't have stack traces for the various calls and I don't really understand where the calls are coming from. But one thing I noticed is a pattern where a couple of endpoints get hit constantly around 2x per second per cgroup, for example:

```
2023-10-12T19:51:27.048190323Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:27.048212364Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:27.048253820Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:27.048282393Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:28.048531428Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:28.048589618Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:28.048808419Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:28.048861550Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:29.048462262Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:29.048489758Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:29.048539508Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:29.048554664Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:30.048867831Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:30.048892779Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:30.048901750Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:30.048904612Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:31.048550803Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:31.048582306Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:31.048590270Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:31.048592914Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:32.048476675Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:32.048647967Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:32.048693285Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:32.048700482Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:33.048543671Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:33.048583382Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:33.048589882Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:33.048592935Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:34.047588103Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:34.047623471Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:34.047643985Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:34.047661280Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:35.048116327Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:35.048140700Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:35.048188511Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:35.048203847Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:36.047705238Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:36.047732007Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:36.047772891Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:36.047785615Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:37.048033569Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:37.048062686Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:37.048152547Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:37.048169799Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:38.047871800Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:38.047896317Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:38.047908652Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:38.047917179Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:39.048458181Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:39.048485296Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:39.048487402Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:39.048494096Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:40.048180815Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:40.048204498Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:40.048261479Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:40.048267484Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:41.047676696Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:41.047703491Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:41.047755819Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:41.047768591Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:42.048059149Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:42.048078615Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622139 subsys=cgroup-manager
2023-10-12T19:51:42.048137692Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
2023-10-12T19:51:42.048150884Z level=debug msg="Pod metadata: {Name:echo-same-node-76847c8f99-ldgbd Namespace:cilium-test IPs:[10.244.3.121 fd00:10:244:3::6d73]}" container-cgroup-id=4622071 subsys=cgroup-manager
```

gandro replied:
The function is called by Hubble every time a `trace-sock` is handled, that might explain the high volume. I did not realize that `GetPodMetadataForContainer` did a channel allocation and communication for each call, that might actually be a performance problem :confused: 

Hubble itself will not create any Hubble flow if the socket is not associated with a pod (i.e. if this function here returns nil), but maybe we should think about also reducing the amount of events emitted by the datapath if that's somehow possible.

aditighag replied:
I vaguely remember having this discussion with Sebastian on whether Hubble cached metadata results. The answer was no as Hubble also retrieves metadata from other subsystems such as ipcache, k8s watcher, etc. Please correct me if that's not the case.

I don't have cycles at the moment, but at some point I would like to revisit this: some of the options being aggregating events at the datapath, or having a cache in the cgroup manager. FWIW, events are being generated at the socket and not packet level, but I agree there is a scope for optimization. Btw the aggregation task is part of the follow-ups meta issue - https://github.com/cilium/cilium/issues/21510.

aditighag replied:
Here is a PR that adds caching - https://github.com/cilium/cilium/pull/32615.
